src/backend/core/services/outline_import.py

```diff
@@ -0,0 +1,205 @@
+
+
+def _ensure_dir_documents(user, dir_path: str, dir_docs: dict[str, models.Document]) -> models.Document | None:
+    """Ensure each path segment in dir_path has a container document.
+
+    Returns the deepest parent document or None when dir_path is empty.
+    """
+    if not dir_path:
+        return None
+
+    parts = [p for p in dir_path.split("/") if p]
+    parent: models.Document | None = None
+    current = ""
+    for part in parts:
+        current = f"{current}/{part}" if current else part
+        if current in dir_docs:
+            parent = dir_docs[current]
+            continue
+
+        if parent is None:
+            doc = models.Document.add_root(
+                depth=1,
+                creator=user,
+                title=part,
+                link_reach=models.LinkReachChoices.RESTRICTED,
+            )
+        else:
+            doc = parent.add_child(creator=user, title=part)
+
+        models.DocumentAccess.objects.update_or_create(
+            document=doc,
+            user=user,
+            defaults={"role": models.RoleChoices.OWNER},
+        )
+        dir_docs[current] = doc
+        parent = doc
+
+    return parent
+
+
+def _upload_attachment(user, doc: models.Document, arcname: str, data: bytes) -> str:
+    """Upload a binary asset into object storage and return its public media URL."""
+    content_type, _ = mimetypes.guess_type(arcname)
+    ext = (arcname.split(".")[-1] or "bin").lower()
+    file_id = uuid.uuid4()
+    key = f"{doc.key_base}/{enums.ATTACHMENTS_FOLDER:s}/{file_id!s}.{ext}"
+    extra_args = {
+        "Metadata": {
+            "owner": str(user.id),
+            "status": enums.DocumentAttachmentStatus.READY,
+        },
+    }
+    if content_type:
+        extra_args["ContentType"] = content_type
+
+    default_storage.connection.meta.client.upload_fileobj(
+        io.BytesIO(data), default_storage.bucket_name, key, ExtraArgs=extra_args
+    )
+    doc.attachments.append(key)
+    doc.save(update_fields=["attachments", "updated_at"])
+    malware_detection.analyse_file(key, document_id=doc.id)
+    return f"{settings.MEDIA_BASE_URL}{settings.MEDIA_URL}{key}"
+
+
+def process_outline_zip(user, zip_bytes: bytes) -> list[str]:
+    """Process an Outline export zip and create Docs documents.
+
+    Returns the list of created document IDs (stringified UUIDs) corresponding to
+    markdown-backed documents. Container folders used to rebuild hierarchy are not listed.
+    """
+    archive = zipfile.ZipFile(io.BytesIO(zip_bytes))
+
+    # Basic Zip Slip protection: refuse absolute or parent-traversal entries
+    for name in archive.namelist():
+        # Normalize to posix separators and check traversal
+        if name.startswith("/") or "\\" in name:
+            raise OutlineImportError("Unsafe path in archive")
+        parts = [p for p in name.split("/") if p]
+        if any(part == ".." for part in parts):
+            raise OutlineImportError("Unsafe path in archive")
```

Comment:
Is this always unsafe or only when `..` goes beyond the root you are iterating over?